Permissions to collection, but not to database, shows QB+SQL on dashboard, but can only view SQL question

**Describe the bug**
There's an inconsistency in how permissions are handled.
When a user only has access to view a collection and no database access, then the user can view a dashboard and it's questions, but can only click to view SQL question. When trying to access QB question, it displays "Sorry, you don’t have permission to see that.", which means that you cannot download data, which was the goal. Reported in the [forum](https://discourse.metabase.com/t/download-the-function-of-the-source-data-table/6765).

**To Reproduce**
1. With admin, create a user in All Users.
2. With admin, in Permissions, no Data Access on database to All Users, and View on Our Analytics to All Users.
3. With admin, create two questions - one QB and one SQL, querying database that user does not have access to. Save the questions in Our Analytics.
4. With admin, create dashboard and add the two questions, save in Our Analytics.
5. With user, view the dashboard (console fetchData error) and it shows both questions.
   - click SQL question (console errors on 0.32.9) and it shows data. (0.33.0-preview2 no console errors, and it shows the SQL query too).
   - click QB question (console errors on 0.32.9) and shows the message "Sorry, you don’t have permission to see that.". (0.33.0-preview2 just shows a spinner and `TypeError: n is null` error in console, so UI stops working)

**Logs**
There's a couple of related console messages, but this is the main one:
```
fetchData error Object { status: 403, data: "You don't have permissions to do that.", isCancelled: false }
```

**Information about your Metabase Installation:**
Metabase 0.32.9 production and 0.33.0-preview2 test, using Sample Dataset, on Firefox.

Reproduced in master. SQL questions do seem to work fine.

If you don't have data access but _do_ have collection access, clicking on a GUI question either from a dashboard as in the example, or from a collection takes you to the simple question pick-your-data step with an infinitely loading spinner.

JS console errors:
```
warning.js:33 Warning: setState(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.
printWarning @ warning.js:33
warning @ warning.js:57
getInternalInstanceReadyForUpdate @ ReactUpdateQueue.js:52
enqueueSetState @ ReactUpdateQueue.js:207
../../../node_modules/react/lib/ReactBaseClasses.js.ReactComponent.setState @ ReactBaseClasses.js:62
_callee$ @ DataSelector.jsx:294
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:318
onBack @ DataSelector.jsx:341
TablePicker @ DataSelector.jsx:781
(anonymous) @ ReactCompositeComponent.js:303
measureLifeCyclePerf @ ReactCompositeComponent.js:73
_constructComponentWithoutOwner @ ReactCompositeComponent.js:302
_constructComponent @ ReactCompositeComponent.js:277
mountComponent @ ReactCompositeComponent.js:185
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
mountComponentIntoNode @ ReactMount.js:102
perform @ Transaction.js:141
batchedMountComponentIntoNode @ ReactMount.js:124
batchedUpdates @ ReactDefaultBatchingStrategy.js:58
batchedUpdates @ ReactUpdates.js:95
_renderNewRootComponent @ ReactMount.js:317
_renderSubtreeIntoContainer @ ReactMount.js:399
renderSubtreeIntoContainer @ ReactMount.js:340
_renderPopover @ Popover.jsx:289
(anonymous) @ makeAssimilatePrototype.js:15
componentDidMount @ Popover.jsx:95
(anonymous) @ makeAssimilatePrototype.js:15
(anonymous) @ ReactCompositeComponent.js:262
measureLifeCyclePerf @ ReactCompositeComponent.js:73
(anonymous) @ ReactCompositeComponent.js:261
notifyAll @ CallbackQueue.js:74
close @ ReactReconcileTransaction.js:78
closeAll @ Transaction.js:207
perform @ Transaction.js:154
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
flushBatchedUpdates @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
batchedUpdates @ ReactUpdates.js:95
dispatchEvent @ ReactEventListener.js:145
Show 32 more frames
warning.js:33 Warning: setState(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.
printWarning @ warning.js:33
warning @ warning.js:57
getInternalInstanceReadyForUpdate @ ReactUpdateQueue.js:52
enqueueSetState @ ReactUpdateQueue.js:207
../../../node_modules/react/lib/ReactBaseClasses.js.ReactComponent.setState @ ReactBaseClasses.js:62
_callee$ @ DataSelector.jsx:294
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:19
(anonymous) @ DataSelector.jsx:318
onBack @ DataSelector.jsx:341
TablePicker @ DataSelector.jsx:781
(anonymous) @ ReactCompositeComponent.js:303
measureLifeCyclePerf @ ReactCompositeComponent.js:73
_constructComponentWithoutOwner @ ReactCompositeComponent.js:302
_constructComponent @ ReactCompositeComponent.js:277
mountComponent @ ReactCompositeComponent.js:185
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
mountChildren @ ReactMultiChild.js:234
_createInitialChildren @ ReactDOMComponent.js:701
mountComponent @ ReactDOMComponent.js:520
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
mountComponentIntoNode @ ReactMount.js:102
perform @ Transaction.js:141
batchedMountComponentIntoNode @ ReactMount.js:124
batchedUpdates @ ReactDefaultBatchingStrategy.js:58
batchedUpdates @ ReactUpdates.js:95
_renderNewRootComponent @ ReactMount.js:317
_renderSubtreeIntoContainer @ ReactMount.js:399
renderSubtreeIntoContainer @ ReactMount.js:340
_renderPopover @ Popover.jsx:289
(anonymous) @ makeAssimilatePrototype.js:15
componentDidMount @ Popover.jsx:95
(anonymous) @ makeAssimilatePrototype.js:15
(anonymous) @ ReactCompositeComponent.js:262
measureLifeCyclePerf @ ReactCompositeComponent.js:73
(anonymous) @ ReactCompositeComponent.js:261
notifyAll @ CallbackQueue.js:74
close @ ReactReconcileTransaction.js:78
closeAll @ Transaction.js:207
perform @ Transaction.js:154
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
flushBatchedUpdates @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
enqueueUpdate @ ReactUpdates.js:198
enqueueUpdate @ ReactUpdateQueue.js:22
enqueueSetState @ ReactUpdateQueue.js:216
../../../node_modules/react/lib/ReactBaseClasses.js.ReactComponent.setState @ ReactBaseClasses.js:62
onStateChange @ connectAdvanced.js:205
notify @ Subscription.js:26
notifyNestedSubs @ Subscription.js:65
onStateChange @ connectAdvanced.js:202
(anonymous) @ makeAssimilatePrototype.js:15
dispatch @ createStore.js:173
(anonymous) @ middleware.js:22
(anonymous) @ index.js:28
(anonymous) @ store.js:17
dispatch @ applyMiddleware.js:35
action @ store.js:22
_callee3$ @ actions.js:445
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ actions.js:77
(anonymous) @ actions.js:77
Promise.then (async)
step @ actions.js:77
(anonymous) @ actions.js:77
Promise.then (async)
step @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:482
(anonymous) @ store.js:16
(anonymous) @ bindActionCreators.js:3
componentWillMount @ QueryBuilder.jsx:170
(anonymous) @ ReactCompositeComponent.js:345
measureLifeCyclePerf @ ReactCompositeComponent.js:73
performInitialMount @ ReactCompositeComponent.js:344
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
_updateRenderedComponent @ ReactCompositeComponent.js:762
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
updateChildren @ ReactChildReconciler.js:107
_reconcilerUpdateChildren @ ReactMultiChild.js:204
_updateChildren @ ReactMultiChild.js:308
updateChildren @ ReactMultiChild.js:295
_updateDOMChildren @ ReactDOMComponent.js:944
updateComponent @ ReactDOMComponent.js:758
receiveComponent @ ReactDOMComponent.js:720
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
performUpdateIfNecessary @ ReactCompositeComponent.js:558
performUpdateIfNecessary @ ReactReconciler.js:154
runBatchedUpdates @ ReactUpdates.js:148
perform @ Transaction.js:141
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
flushBatchedUpdates @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
batchedUpdates @ ReactUpdates.js:95
dispatchEvent @ ReactEventListener.js:145
Show 150 more frames
actions.js:529 Error loading metadata for card TypeError: Cannot read property 'fields' of null
    at StructuredQuery.dependentTableIds (StructuredQuery.js:1370)
    at _callee5$ (actions.js:529)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:271)
    at Generator.prototype.<computed> [as next] (runtime.js:97)
    at step (actions.js:77)
    at actions.js:77
    at new Promise (<anonymous>)
    at actions.js:77
    at actions.js:556
_callee5$ @ actions.js:529
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:556
_callee3$ @ redux.js:205
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ redux.js:15
(anonymous) @ redux.js:15
(anonymous) @ redux.js:15
(anonymous) @ redux.js:215
(anonymous) @ store.js:16
dispatch @ applyMiddleware.js:35
_callee3$ @ actions.js:459
tryCatch @ runtime.js:45
invoke @ runtime.js:271
prototype.<computed> @ runtime.js:97
step @ actions.js:77
(anonymous) @ actions.js:77
Promise.then (async)
step @ actions.js:77
(anonymous) @ actions.js:77
Promise.then (async)
step @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:77
(anonymous) @ actions.js:482
(anonymous) @ store.js:16
(anonymous) @ bindActionCreators.js:3
componentWillMount @ QueryBuilder.jsx:170
(anonymous) @ ReactCompositeComponent.js:345
measureLifeCyclePerf @ ReactCompositeComponent.js:73
performInitialMount @ ReactCompositeComponent.js:344
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
performInitialMount @ ReactCompositeComponent.js:368
mountComponent @ ReactCompositeComponent.js:255
mountComponent @ ReactReconciler.js:43
_updateRenderedComponent @ ReactCompositeComponent.js:762
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
updateChildren @ ReactChildReconciler.js:107
_reconcilerUpdateChildren @ ReactMultiChild.js:204
_updateChildren @ ReactMultiChild.js:308
updateChildren @ ReactMultiChild.js:295
_updateDOMChildren @ ReactDOMComponent.js:944
updateComponent @ ReactDOMComponent.js:758
receiveComponent @ ReactDOMComponent.js:720
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
receiveComponent @ ReactCompositeComponent.js:544
receiveComponent @ ReactReconciler.js:122
_updateRenderedComponent @ ReactCompositeComponent.js:751
_performComponentUpdate @ ReactCompositeComponent.js:721
updateComponent @ ReactCompositeComponent.js:642
performUpdateIfNecessary @ ReactCompositeComponent.js:558
performUpdateIfNecessary @ ReactReconciler.js:154
runBatchedUpdates @ ReactUpdates.js:148
perform @ Transaction.js:141
perform @ Transaction.js:141
perform @ ReactUpdates.js:87
flushBatchedUpdates @ ReactUpdates.js:170
closeAll @ Transaction.js:207
perform @ Transaction.js:154
batchedUpdates @ ReactDefaultBatchingStrategy.js:60
batchedUpdates @ ReactUpdates.js:95
dispatchEvent @ ReactEventListener.js:145
Show 95 more frames
StructuredQuery.js:1370 Uncaught (in promise) TypeError: Cannot read property 'fields' of null
    at StructuredQuery.dependentTableIds (StructuredQuery.js:1370)
    at _callee5$ (actions.js:529)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:271)
    at Generator.prototype.<computed> [as next] (runtime.js:97)
    at step (actions.js:77)
    at actions.js:77
    at new Promise (<anonymous>)
    at actions.js:77
    at actions.js:556
```

If I click on a GUI question from a dashboard, I do see a 403 in the server logs, but I don't see any errors if I click on a GUI question from a collection.
```
08-01 10:56:09 WARN middleware.log :: GET /api/database/1/metadata 403 7 ms (3 DB calls) 
"You don't have permissions to do that."

08-01 10:56:09 INFO api.card :: Question's average execution duration is 74 ms; using 'magic' TTL of 740 seconds 💾
08-01 10:56:09 INFO api.card :: Question's average execution duration is 67 ms; using 'magic' TTL of 670 seconds 💾
08-01 10:56:09 INFO middleware.cache :: Query took 49 ms to run; miminum for cache eligibility is 1000 ms
08-01 10:56:09 DEBUG middleware.log :: POST /api/card/292/query 200 [ASYNC: completed] 70 ms (12 DB calls) Jetty threads: 4/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 1
08-01 10:56:09 INFO middleware.cache :: Query took 58 ms to run; miminum for cache eligibility is 1000 ms
08-01 10:56:09 DEBUG middleware.log :: POST /api/card/291/query 200 [ASYNC: completed] 79 ms (18 DB calls) Jetty threads: 4/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 1
08-01 10:56:16 DEBUG middleware.log :: GET /api/session/properties 200 4 ms (1 DB calls) Jetty threads: 6/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
08-01 10:56:16 DEBUG middleware.log :: GET /api/user/current 200 6 ms (3 DB calls) Jetty threads: 6/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
08-01 10:56:16 DEBUG middleware.log :: GET /api/database 200 3 ms (3 DB calls) Jetty threads: 5/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
08-01 10:56:16 DEBUG middleware.log :: GET /api/database 200 3 ms (3 DB calls) Jetty threads: 5/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
08-01 10:56:16 DEBUG middleware.log :: GET /api/card/291 200 3 ms (5 DB calls) Jetty threads: 5/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
08-01 10:56:16 DEBUG middleware.log :: GET /api/alert/question/291 200 1 ms (1 DB calls) Jetty threads: 5/50 (2 idle, 0 queued) (111 total active threads) Queries in flight: 0
```

--- a/frontend/src/metabase-lib/lib/queries/StructuredQuery.js
+++ b/frontend/src/metabase-lib/lib/queries/StructuredQuery.js
@@ -107,15 +107,21 @@ export default class StructuredQuery extends AtomicQuery {
    * @returns true if this query is in a state where it can be run.
    */
   canRun() {
-    // TODO: do we need other logic here or do we now ensure queries are always valid?
-    return !!this.table();
+    return !!(this.sourceTableId() || this.sourceQuery());
+  }
+
+  /**
+   * @returns true if we have metadata for the root source table loaded
+   */
+  hasMetadata() {
+    return this.metadata() && !!this.rootTable();
   }
 
   /**
    * @returns true if this query is in a state where it can be edited. Must have database and table set, and metadata for the table loaded.
    */
   isEditable(): boolean {
-    return !!this.tableMetadata();
+    return this.hasMetadata();
   }
 
   /* AtomicQuery superclass methods */
@@ -208,17 +214,16 @@ export default class StructuredQuery extends AtomicQuery {
   }
 
   /**
-   * @returns a new query with the provided Table set.
+   * @returns the table ID, if a table is selected.
    */
-  setTable(table: Table): StructuredQuery {
-    return this.setTableId(table.id);
+  sourceTableId(): ?TableId {
+    return this.query()["source-table"];
   }
-
   /**
    * @returns a new query with the provided Table ID set.
    */
-  setTableId(tableId: TableId): StructuredQuery {
-    if (tableId !== this.tableId()) {
+  setSourceTableId(tableId: TableId): StructuredQuery {
+    if (tableId !== this.sourceTableId()) {
       return new StructuredQuery(
         this._originalQuestion,
         chain(this.datasetQuery())
@@ -232,10 +237,22 @@ export default class StructuredQuery extends AtomicQuery {
   }
 
   /**
-   * @returns the table ID, if a table is selected.
+   * @deprecated: use sourceTableId
    */
   tableId(): ?TableId {
-    return this.query()["source-table"];
+    return this.sourceTableId();
+  }
+  /**
+   * @deprecated: use setSourceTableId
+   */
+  setTableId(tableId: TableId): StructuredQuery {
+    return this.setSourceTableId(tableId);
+  }
+  /**
+   * @deprecated: use setSourceTableId
+   */
+  setTable(table: Table): StructuredQuery {
+    return this.setSourceTableId(table.id);
   }
 
   /**
@@ -296,7 +313,7 @@ export default class StructuredQuery extends AtomicQuery {
       augmentDatabase({ tables: [table] });
       return table;
     } else {
-      return this.metadata().table(this.tableId());
+      return this.metadata().table(this.sourceTableId());
     }
   }
 
@@ -318,7 +335,7 @@ export default class StructuredQuery extends AtomicQuery {
    * Removes invalid clauses from the query (and source-query, recursively)
    */
   clean() {
-    if (!this.metadata()) {
+    if (!this.hasMetadata()) {
       console.warn("Warning: can't clean query without metadata!");
       return this;
     }
@@ -1381,9 +1398,8 @@ export default class StructuredQuery extends AtomicQuery {
 
   /**
    * returns the original Table object at the beginning of the nested queries
-   * NOTE: this is inconsistent with sourceQuery() returning the `source-query`. Should we swap `table()` and `sourceTable()`?
    */
-  sourceTable(): Table {
+  rootTable(): Table {
     return this.rootQuery().table();
   }
 
@@ -1415,15 +1431,17 @@ export default class StructuredQuery extends AtomicQuery {
     const tableIds = new Set();
 
     // source-table, if set
-    const tableId = this.tableId();
+    const tableId = this.sourceTableId();
     if (tableId) {
       tableIds.add(tableId);
       // implicit joins via foreign keys
       if (includeFKs) {
         const table = this.table();
-        for (const field of table.fields) {
-          if (field.target && field.target.table_id) {
-            tableIds.add(field.target.table_id);
+        if (table) {
+          for (const field of table.fields) {
+            if (field.target && field.target.table_id) {
+              tableIds.add(field.target.table_id);
+            }
           }
         }
       }

--- a/frontend/src/metabase/query_builder/actions.js
+++ b/frontend/src/metabase/query_builder/actions.js
@@ -534,11 +534,11 @@ export const loadMetadataForCard = createThunkAction(
       const query = new Question(getMetadata(getState()), card).query();
       if (query instanceof StructuredQuery) {
         try {
-          const sourceTable = query.sourceTable();
-          if (sourceTable) {
+          const rootTable = query.rootTable();
+          if (rootTable) {
             await Promise.all([
-              dispatch(Tables.actions.fetchTableMetadata(sourceTable)),
-              dispatch(Tables.actions.fetchForeignKeys(sourceTable)),
+              dispatch(Tables.actions.fetchTableMetadata(rootTable)),
+              dispatch(Tables.actions.fetchForeignKeys(rootTable)),
             ]);
           }
           await Promise.all(
@@ -761,22 +761,26 @@ export const updateQuestion = (
       dispatch(setIsShowingTemplateTagsEditor(false));
     }
 
-    if (
-      !_.isEqual(
-        oldQuestion.query().dependentTableIds(),
-        newQuestion.query().dependentTableIds(),
-      )
-    ) {
-      await dispatch(loadMetadataForCard(newQuestion.card()));
-    }
+    try {
+      if (
+        !_.isEqual(
+          oldQuestion.query().dependentTableIds(),
+          newQuestion.query().dependentTableIds(),
+        )
+      ) {
+        await dispatch(loadMetadataForCard(newQuestion.card()));
+      }
 
-    // setDefaultQuery requires metadata be loaded, need getQuestion to use new metadata
-    const question = getQuestion(getState());
-    const questionWithDefaultQuery = question.setDefaultQuery();
-    if (!questionWithDefaultQuery.isEqual(question)) {
-      await dispatch.action(UPDATE_QUESTION, {
-        card: questionWithDefaultQuery.setDefaultDisplay().card(),
-      });
+      // setDefaultQuery requires metadata be loaded, need getQuestion to use new metadata
+      const question = getQuestion(getState());
+      const questionWithDefaultQuery = question.setDefaultQuery();
+      if (!questionWithDefaultQuery.isEqual(question)) {
+        await dispatch.action(UPDATE_QUESTION, {
+          card: questionWithDefaultQuery.setDefaultDisplay().card(),
+        });
+      }
+    } catch (e) {
+      // this will fail if user doesn't have data permissions but thats ok
     }
 
     // run updated query

--- a/frontend/src/metabase/query_builder/components/notebook/steps/DataStep.jsx
+++ b/frontend/src/metabase/query_builder/components/notebook/steps/DataStep.jsx
@@ -8,6 +8,7 @@ import { NotebookCell, NotebookCellItem } from "../NotebookCell";
 import { getDatabasesList } from "metabase/query_builder/selectors";
 
 function DataStep({ color, query, databases, updateQuery }) {
+  const table = query.table();
   return (
     <NotebookCell color={color}>
       <DatabaseSchemaAndTableDataSelector
@@ -28,12 +29,12 @@ function DataStep({ color, query, databases, updateQuery }) {
             </NotebookCellItem>
           ) : (
             <NotebookCellItem color={color} icon="table2">
-              {query.table().displayName()}
+              {table && table.displayName()}
             </NotebookCellItem>
           )
         }
       />
-      {query.table() && query.isRaw() && (
+      {table && query.isRaw() && (
         <DataFieldsPicker
           className="ml-auto mb1 text-bold"
           query={query}

--- a/frontend/src/metabase/query_builder/components/view/QuestionDataSource.jsx
+++ b/frontend/src/metabase/query_builder/components/view/QuestionDataSource.jsx
@@ -8,15 +8,26 @@ import StructuredQuery from "metabase-lib/lib/queries/StructuredQuery";
 
 import cx from "classnames";
 
-const QuestionDataSource = ({
-  question,
-  query = question.query(),
-  subHead,
-  noLink,
-  ...props
-}) => {
+const QuestionDataSource = ({ question, subHead, noLink, ...props }) => {
+  const parts = getDataSourceParts({ question, subHead, noLink });
+  return subHead ? (
+    <SubHeadBreadcrumbs parts={parts} {...props} />
+  ) : (
+    <HeadBreadcrumbs parts={parts} {...props} />
+  );
+};
+
+QuestionDataSource.shouldRender = ({ question }) =>
+  getDataSourceParts({ question }).length > 0;
+
+function getDataSourceParts({ question, noLink, subHead }) {
+  if (!question) {
+    return [];
+  }
+
   const parts = [];
 
+  let query = question.query();
   if (query instanceof StructuredQuery) {
     query = query.rootQuery();
   }
@@ -67,12 +78,8 @@ const QuestionDataSource = ({
     });
   }
 
-  return subHead ? (
-    <SubHeadBreadcrumbs parts={parts} {...props} />
-  ) : (
-    <HeadBreadcrumbs parts={parts} {...props} />
-  );
-};
+  return parts.filter(({ name, icon }) => name || icon);
+}
 
 export default QuestionDataSource;
 

--- a/frontend/src/metabase/query_builder/components/view/QuestionEntityMenu.jsx
+++ b/frontend/src/metabase/query_builder/components/view/QuestionEntityMenu.jsx
@@ -5,38 +5,43 @@ import EntityMenu from "metabase/components/EntityMenu";
 
 import cx from "classnames";
 
-const QuestionEntityMenu = ({ className, onOpenModal }) => (
-  <EntityMenu
-    triggerIcon="pencil"
-    className={cx("text-light", className)}
-    items={[
-      {
-        icon: "edit_document",
-        title: t`Edit this question`,
-        action: () => onOpenModal("edit"),
-      },
-      {
-        icon: "history",
-        title: t`View revision history`,
-        action: () => onOpenModal("history"),
-      },
-      {
-        icon: "add_to_dash",
-        title: t`Add to dashboard`,
-        action: () => onOpenModal("add-to-dashboard"),
-      },
-      {
-        icon: "move",
-        title: t`Move`,
-        action: () => onOpenModal("move"),
-      },
-      {
-        icon: "archive",
-        title: `Archive`,
-        action: () => onOpenModal("archive"),
-      },
-    ]}
-  />
-);
-
-export default QuestionEntityMenu;
+export default function QuestionEntityMenu({
+  className,
+  question,
+  onOpenModal,
+}) {
+  const canWrite = question && question.canWrite();
+  return (
+    <EntityMenu
+      triggerIcon="pencil"
+      className={cx("text-light", className)}
+      items={[
+        canWrite && {
+          icon: "edit_document",
+          title: t`Edit this question`,
+          action: () => onOpenModal("edit"),
+        },
+        {
+          icon: "history",
+          title: t`View revision history`,
+          action: () => onOpenModal("history"),
+        },
+        {
+          icon: "add_to_dash",
+          title: t`Add to dashboard`,
+          action: () => onOpenModal("add-to-dashboard"),
+        },
+        canWrite && {
+          icon: "move",
+          title: t`Move`,
+          action: () => onOpenModal("move"),
+        },
+        canWrite && {
+          icon: "archive",
+          title: `Archive`,
+          action: () => onOpenModal("archive"),
+        },
+      ]}
+    />
+  );
+}

--- a/frontend/src/metabase/query_builder/components/view/QuestionFilters.jsx
+++ b/frontend/src/metabase/query_builder/components/view/QuestionFilters.jsx
@@ -98,11 +98,12 @@ export function QuestionFilterWidget({
 QuestionFilters.shouldRender = ({ question, queryBuilderMode }) =>
   queryBuilderMode === "view" &&
   question.isStructured() &&
+  question.query().isEditable() &&
   question.query().topLevelFilters().length > 0 &&
   !question.isObjectDetail();
 
 QuestionFilterWidget.shouldRender = ({ question, queryBuilderMode }) =>
   queryBuilderMode === "view" &&
   question.isStructured() &&
-  question.query().table() &&
+  question.query().isEditable() &&
   !question.isObjectDetail();

--- a/frontend/src/metabase/query_builder/components/view/QuestionNotebookButton.jsx
+++ b/frontend/src/metabase/query_builder/components/view/QuestionNotebookButton.jsx
@@ -1,0 +1,36 @@
+import React from "react";
+
+import { t } from "ttag";
+import cx from "classnames";
+
+import Tooltip from "metabase/components/Tooltip";
+import Button from "metabase/components/Button";
+
+export default function QuestionNotebookButton({
+  className,
+  question,
+  isShowingNotebook,
+  setQueryBuilderMode,
+  ...props,
+}) {
+  return QuestionNotebookButton.shouldRender({ question }) ? (
+    <Tooltip tooltip={isShowingNotebook ? t`Hide editor` : t`Show editor`}>
+      <Button
+        borderless={!isShowingNotebook}
+        primary={isShowingNotebook}
+        medium
+        className={cx(className, {
+          "text-brand-hover": !isShowingNotebook,
+        })}
+        icon="notebook"
+        onClick={() =>
+          setQueryBuilderMode(isShowingNotebook ? "view" : "notebook")
+        }
+        {...props}
+      />
+    </Tooltip>
+  ) : null;
+}
+
+QuestionNotebookButton.shouldRender = ({ question }) =>
+  question.isStructured() && question.query().isEditable();

--- a/frontend/src/metabase/query_builder/components/view/QuestionNotebookButton.jsx
+++ b/frontend/src/metabase/query_builder/components/view/QuestionNotebookButton.jsx
@@ -11,7 +11,7 @@ export default function QuestionNotebookButton({
   question,
   isShowingNotebook,
   setQueryBuilderMode,
-  ...props,
+  ...props
 }) {
   return QuestionNotebookButton.shouldRender({ question }) ? (
     <Tooltip tooltip={isShowingNotebook ? t`Hide editor` : t`Show editor`}>

--- a/frontend/src/metabase/query_builder/components/view/View.jsx
+++ b/frontend/src/metabase/query_builder/components/view/View.jsx
@@ -113,7 +113,12 @@ export default class View extends React.Component {
     const isStructured = query instanceof StructuredQuery;
     const isNative = query instanceof NativeQuery;
 
-    if (isStructured && queryBuilderMode === "view" && !query.table()) {
+    const isNewQuestion =
+      query instanceof StructuredQuery &&
+      !query.sourceTableId() &&
+      !query.sourceQuery();
+
+    if (isNewQuestion && queryBuilderMode === "view") {
       return (
         <div className={fitClassNames}>
           <div className="p4 mx2">
@@ -168,8 +173,6 @@ export default class View extends React.Component {
           onClose={() => this.props.toggleDataReference()}
         />
       ) : null;
-
-    const isNewQuestion = query instanceof StructuredQuery && !query.table();
 
     const isSidebarOpen = leftSideBar || rightSideBar;
 

--- a/frontend/src/metabase/query_builder/components/view/ViewHeader.jsx
+++ b/frontend/src/metabase/query_builder/components/view/ViewHeader.jsx
@@ -5,10 +5,8 @@ import { Box } from "grid-styled";
 
 import Icon from "metabase/components/Icon";
 import Link from "metabase/components/Link";
-import Button from "metabase/components/Button";
 import ButtonBar from "metabase/components/ButtonBar";
 import CollectionBadge from "metabase/questions/components/CollectionBadge";
-import Tooltip from "metabase/components/Tooltip.jsx";
 
 import ViewSection, { ViewHeading, ViewSubHeading } from "./ViewSection";
 
@@ -17,6 +15,7 @@ import QuestionDescription from "./QuestionDescription";
 import QuestionEntityMenu from "./QuestionEntityMenu";
 import QuestionLineage from "./QuestionLineage";
 import QuestionPreviewToggle from "./QuestionPreviewToggle";
+import QuestionNotebookButton from "./QuestionNotebookButton";
 
 import QuestionFilters, { QuestionFilterWidget } from "./QuestionFilters";
 import { QuestionSummarizeWidget } from "./QuestionSummaries";
@@ -126,9 +125,17 @@ export class ViewTitleHeader extends React.Component {
                 collectionId={question.collectionId()}
               />
 
-              <span className="mb1 mx2 text-light text-smaller">•</span>
+              {QuestionDataSource.shouldRender({ question }) && (
+                <span className="mb1 mx2 text-light text-smaller">•</span>
+              )}
 
-              <QuestionDataSource className="mb1" question={question} subHead />
+              {QuestionDataSource.shouldRender({ question }) && (
+                <QuestionDataSource
+                  className="mb1"
+                  question={question}
+                  subHead
+                />
+              )}
 
               {QuestionFilters.shouldRender(this.props) && (
                 <QuestionFilters
@@ -217,24 +224,14 @@ export class ViewTitleHeader extends React.Component {
               onCloseSummary={onCloseSummary}
             />
           )}
-          {question.isStructured() && (
-            <Tooltip
-              tooltip={isShowingNotebook ? t`Hide editor` : t`Show editor`}
-            >
-              <Button
-                borderless={!isShowingNotebook}
-                primary={isShowingNotebook}
-                medium
-                className={cx("hide sm-show", {
-                  "text-brand-hover": !isShowingNotebook,
-                })}
-                icon="notebook"
-                ml={2}
-                onClick={() =>
-                  setQueryBuilderMode(isShowingNotebook ? "view" : "notebook")
-                }
-              />
-            </Tooltip>
+          {QuestionNotebookButton.shouldRender({ question }) && (
+            <QuestionNotebookButton
+              className="hide sm-show"
+              ml={2}
+              question={question}
+              isShowingNotebook={isShowingNotebook}
+              setQueryBuilderMode={setQueryBuilderMode}
+            />
           )}
           {NativeQueryButton.shouldRender(this.props) && (
             <Box

